src/Middleware/StaticFiles/src/StaticFilesEndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,223 @@
+        /// will be used.
+        /// </para>
+        /// <para>
+        /// The order of the registered endpoint will be <c>int.MaxValue</c>.
+        /// </para>
+        /// <para>
+        /// This overload will use the provided <paramref name="pattern"/> verbatim. Use the <c>:nonfile</c> route contraint
+        /// to exclude requests for static files.
+        /// </para>
+        /// </remarks>
+        public static IEndpointConventionBuilder MapFallbackToFile(
+            this IEndpointRouteBuilder builder,
+            string pattern,
+            string filePath)
+        {
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            if (pattern == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            if (filePath == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            return builder.MapFallback(pattern, CreateRequestDelegate(builder, filePath));
+        }
+
+        /// <summary>
+        /// Registers a fallback <see cref="RouteEndpoint"/> that will serve the provided static file
+        /// </summary>
+        /// <param name="builder">The <see cref="IEndpointRouteBuilder"/>.</param>\
+        /// <param name="pattern">The route pattern.</param>
+        /// <param name="filePath">The file path of the file to serve.</param>
+        /// <param name="options"><see cref="StaticFileOptions"/> for the <see cref="StaticFileMiddleware"/>.</param>
+        /// <returns>The <see cref="IEndpointRouteBuilder"/></returns>
+        /// <remarks>
+        /// <para>
+        /// <see cref="MapFallbackToFile(IEndpointRouteBuilder, string)"/> is intended to handle cases where URL path of
+        /// the request does not contain a filename, and no other endpoint has matched. This is convenient for routing
+        /// requests for dynamic content to a SPA framework, while also allowing requests for non-existent files to
+        /// result in an HTTP 404.
+        /// </para>
+        /// <para>
+        /// <see cref="MapFallbackToFile(IEndpointRouteBuilder, string)"/> with use the <see cref="StaticFileMiddleware"/>
+        /// to serve requests for the provided <paramref name="filePath"/>.
+        /// </para>
+        /// <para>
+        /// The order of the registered endpoint will be <c>int.MaxValue</c>.
+        /// </para>
+        /// <para>
+        /// This overload will use the provided <paramref name="pattern"/> verbatim. Use the <c>:nonfile</c> route contraint
+        /// to exclude requests for static files.
+        /// </para>
+        /// </remarks>
+        public static IEndpointConventionBuilder MapFallbackToFile(
+            this IEndpointRouteBuilder builder,
+            string pattern,
+            string filePath,
+            StaticFileOptions options)
```

Comment:
@SteveSandersonMS - I carried forward the options from the spa services stuff because it seemed like there was a lot of complexity there that I don't totally get.

I'm interested in whether you think these things can be used to solve the same set of problems or not.

Reply:
I would strongly expect that `MapFallbackToFile` could replace the bulk of `UseSpaStaticFiles`.

The `StaticFileOptions` are needed if you want to set custom response headers (e.g., for caching policy), which people do.

Other complications in `UseSpaStaticFiles` are to do with things like "at dev time I proxy requests to a SPA dev server, but in production is serve prebuilt static files from my contentroot".

Reply:
OK cool. That was what I was hoping for. I'm confident that we can build more flavours of this if needed.